packages/utils/telemetry-utils/src/errorLogging.ts

```diff
@@ -63,31 +70,121 @@ const isRwLoggingError = (x: any): x is RwLoggingError =>
+ * @param newErrorFn - callback that will create a new error given the original error's message
+ * @returns A new error object "wrapping" the given error
+ */
+ export function wrapError<T>(
+    error: any,
+    newErrorFn: (m: string) => T,
+): T {
+    const {
+        message,
+        stack,
+    } = extractLogSafeErrorProperties(error);
+    const props = isILoggingError(error) ? error.getTelemetryProperties() : {};
+
+    const newError = newErrorFn(message);
+    mixinTelemetryProps(newError as ExtensibleObject<T>, props);
+
+    if (stack !== undefined) {
+        Object.assign(newError, { stack });
+    }
+
+    return newError;
+}
+
+/**
+ * Normalize the given error object yielding a valid Fluid Error
+ * @returns The same error object passed in if possible, normalized and with any provided annotations applied
+ */
+export function normalizeError(
+    error: unknown,
+    annotations: {
+        props?: ITelemetryProperties,
+        errorCodeIfNone?: string,
+    } = {},
+): IFluidErrorBase {
+    // Set up some helpers
+    function setErrorTypeIfMissing(builder: Builder<IFluidErrorBase>, errorTypeIfNone: string) {
+        if (!hasErrorType(error)) {
+            builder.errorType = `none (${errorTypeIfNone})`;
+        }
+    }
+    const fullErrorCodeifNone = annotations.errorCodeIfNone === undefined
+        ? "none"
+        : `none (${annotations.errorCodeIfNone})`;
+
+    // We'll be sure to set all properties on here before casting to IFluidErrorBase and returning
+    let fluidErrorBuilder: Builder<IFluidErrorBase>;
+
+    // If we can't annotate the error, wrap it
+    if (!isRegularObject(error) || Object.isFrozen(error)) {
+        const errorType = !isRegularObject(error) ? typeof(error) : "wrappedFrozenError";
+        const newErrorFn = (errMsg: string) => {
+            fluidErrorBuilder = new LoggingError(errMsg, annotations.props) as Builder<IFluidErrorBase>;
+            setErrorTypeIfMissing(fluidErrorBuilder, errorType);
+            fluidErrorBuilder.fluidErrorCode = fullErrorCodeifNone;
+            return fluidErrorBuilder as IFluidErrorBase;
+        };
+        return wrapError<IFluidErrorBase>(error, newErrorFn);
+    }
+
+    // Do we already have a valid Fluid Error?  Then just mixin telemetry props
+    if (isFluidError(error)) {
+        return annotations.props !== undefined
+            ? mixinTelemetryProps(error as ExtensibleObject<IFluidErrorBase>, annotations.props)
+            : error;
+    }
+
+    // We have a mutable object, not already a valid Fluid Error. Time to fill in the gaps!
+    fluidErrorBuilder = error as Builder<IFluidErrorBase>;
+    fluidErrorBuilder.fluidErrorCode = fullErrorCodeifNone;
+
+    if (isErrorLike(error)) {
+        setErrorTypeIfMissing(fluidErrorBuilder, error.name);
+    } else {
+        setErrorTypeIfMissing(fluidErrorBuilder, typeof(error));
+        fluidErrorBuilder.message = annotations.errorCodeIfNone;
+        fluidErrorBuilder.name = "none";
```

Comment:
its cheaper to store undefined. what's the advantage of setting name to `"none"`

Reply:
You had me convinced, until I realized that my type guard depends on this property being present.  The definitive presence of this property is a nice way to distinguish both at runtime and in the logs which errors adhere to the new type.

Reply:
Well, the type extends `Error` today, where `name` is required.  Maybe it works to extend `Partial<Error>`, I'll think about it.  You know how I feel about `message` ;P  and `stack` is already optional.

Reply:
Done, I went the `Partial<Error>` route, good feedback.